Document/0x05g-Testing-Network-Communication.md

```diff
@@ -1,4 +1,130 @@
-# Android Network APIs
+# Android Network Communication
+
+Almost every Android app acts as a client to one or more remote services. As this network communication usually takes place over untrusted networks such as public Wi-Fi, classical network based-attacks become a potential issue.
+
+Most modern mobile apps use variants of HTTP-based web services, as these protocols are well-documented and supported.
+
+## Overview
+
+### Android Network Security Configuration
+
+Starting on Android 7.0 (API level 24), Android apps can customize their network security settings using the so-called [Network Security Configuration](https://developer.android.com/training/articles/security-config) feature which offers the following key capabilities:
+
+- **Cleartext traffic**: Protect apps from accidental usage of cleartext traffic (or enables it).
+- **Custom trust anchors**: Customize which Certificate Authorities (CA) are trusted for an app's secure connections. For example, trusting particular self-signed certificates or restricting the set of public CAs that the app trusts.
+- **Certificate pinning**: Restrict an app's secure connection to particular certificates.
+- **Debug-only overrides**: Safely debug secure connections in an app without added risk to the installed base.
+
+If an app defines a custom Network Security Configuration, you can obtain its location by searching for `android:networkSecurityConfig` in the AndroidManifest.xml file.
+
+```xml
+<application android:networkSecurityConfig="@xml/network_security_config"
+```
+
+In this case the file is located at `@xml` (equivalent to /res/xml) and has the name "network_security_config" (which might vary). You should be able to find it as "res/xml/network_security_config.xml". If a configuration exists, the following event should be visible in the [system logs](0x05b-Basic-Security_Testing.md#monitoring-system-logs):
+
+```bash
+D/NetworkSecurityConfig: Using Network Security Config from resource network_security_config
+```
+
+The Network Security Configuration is [XML-based](https://developer.android.com/training/articles/security-config#FileFormat) and can be used to configure app-wide and domain-specific settings:
+
+- `base-config` applies to all connections that the app attempts to make.
+- `domain-config` overrides `base-config` for specific domains (it can contain multiple `domain` entries).
+
+For example, the following configuration uses the `base-config` to prevent cleartext traffic for all domains. But it overrides that rule using a `domain-config`, explicitly allowing cleartext traffic for `localhost` and its subdomains.
```

Comment:
```suggestion
For example, the following configuration uses the `base-config` to prevent cleartext traffic for all domains. But it overrides that rule using a `domain-config`, explicitly allowing cleartext traffic for `localhost`.
```